File: linda-logic/src/main/java/it/unibo/coordination/linda/logic/LogicTemplateImpl.java
package it.unibo.coordination.linda.logic;

import alice.tuprolog.Prolog;
import alice.tuprolog.SolveInfo;
import alice.tuprolog.Struct;
import alice.tuprolog.Term;
import it.unibo.coordination.linda.core.Tuple;
import it.unibo.coordination.prologx.PrologUtils;

import java.util.Objects;

final class LogicTemplateImpl implements LogicTemplate {

    private static final Prolog ENGINE = new Prolog();
    private final Struct term;

    LogicTemplateImpl(final Term term) {
        Objects.requireNonNull(term);

        if (term instanceof Struct && LogicTemplate.getPattern().match(term)) {
            this.term = (Struct) term;
        } else {
            this.term = LogicTemplate.getPattern(term);
        }
    }

    @Override
    public LogicMatch matchWith(Tuple tuple) {
        if (tuple instanceof LogicTuple) {
            final LogicTuple logicTuple = (LogicTuple) tuple;
            final SolveInfo si = ENGINE.solve(PrologUtils.unificationTerm(getTemplate(), logicTuple.getTuple()));
            return new LogicMatchImpl(this, si, logicTuple);
        }

        return new LogicMatchImpl(this, null, null);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        LogicTemplate that = (LogicTemplate) o;
        return Objects.equals(term, that.asTerm());
    }

    @Override
    public int hashCode() {
        return Objects.hash(term.toString());
    }

    @Override
    public String toString() {
        return term.toString();
    }

    public Struct asTerm() {
        return term;
    }

    @Override
    public Term getTemplate() {
        return term.getArg(0);
    }

}

File: linda-logic/src/main/kotlin/it/unibo/coordination/linda/logic/LogicTemplateImpl.kt
package it.unibo.coordination.linda.logic

import it.unibo.tuprolog.core.Struct
import it.unibo.tuprolog.core.Term
import it.unibo.tuprolog.unify.Unificator.Companion.matches

internal class LogicTemplateImpl(term: Term) : LogicTemplate {

    private val term: Struct by lazy {
        if (term is Struct && LogicTemplate.pattern.matches(term)) {
            term
        } else {
            LogicTemplate.getPattern(term)
        }
    }

    override val template: Term
        get() = term[0]

    override fun matchWith(tuple: LogicTuple): LogicMatch = LogicMatchImpl(this, tuple)

    override fun equals(other: Any?): Boolean =
            other is LogicTemplate && LogicTemplate.equals(this, other as LogicTemplate?)

    override fun hashCode(): Int = LogicTemplate.hashCode(this)

    override fun toString(): String = term.toString()

    override fun asTerm(): Struct = term
}
